// gecoviz.js/src/popper.js
import { scaleOrdinal, select } from 'd3';
import $ from 'jquery';
import { createPopper } from '@popperjs/core';
import { colors269 } from './colors269';
import protDomains from './domains';
import {
    capitalize,
    cleanString,
    nonEmptyArray,
} from './helpers';

var PopperCreate = function(selector, d, URLs) {
    function get_PopperHTML(d) {
        let arrayData = []
        let showFields = [
            'gene name',
            'gene',
            'description',
            'anchor',
            'pos',
            'start',
            'end',
            'size',
            'strand'
        ]
        let hideFields = [
            'vStart',
            'vEnd',
            'vSize',
            'geneWidth'
        ]
        Object.entries(d).forEach(([key, field]) => {
            let fieldData = "";
            if (nonEmptyArray(field)) {
                fieldData = '<ul class="popper-ul">\
                    <li class="popper-ul-title">'
                    + key.toUpperCase()
                    + '</li>';
                field.forEach(f => {
                    fieldData += '<li>';
                    fieldData += !URLs[key]
                    ? `<em>${f.id}</em>`
                    : '<a href="'
                        + URLs[key].b
                        + String(f.id)
                        + URLs[key].a
                        + '" target="_blank" style="outline:none;">'
                        + String(f.id)+'</a>';
                    let levelData = !f.level
                            ? ''
                            : f.leveDesc
                            ? ' (level: '
                                + f.level
                                + ', description: '
                                + d.levelDesc
                                + ')'
                            : ' (level: '
                                + f.level
                                + ')';
                    fieldData += levelData
                    fieldData += '<br>' + (f.description || '') + '</li>';
                })
                fieldData += '</ul>';
            } else {
                if (typeof field != 'object'
                && !showFields.includes(key)
                && !hideFields.includes(key)) showFields.push(key)
            }
            if (fieldData) arrayData.push(fieldData)
        })
        var popperHTML = ''; //<strong>Gene information</strong>
        popperHTML += '<div class="p-2">';
        showFields.forEach(f => {
            if (d[f]) popperHTML += `${capitalize(f)}: ${d[f]}<br>`;
        })
        popperHTML += '</div>';
        if (nonEmptyArray(d.pfam)) {
             let dom_id = 'dom' + cleanString(d.anchor + d.pos);
             popperHTML += '<div class="py-2" id=' + dom_id + '></div>'
        }
        if (arrayData.length > 0) popperHTML +=
            '<div class="popper-uls">'
                + arrayData.reduce((t, d) => t + d)
                +'</div>';
        return popperHTML
    }

    var geneID = cleanString(d.anchor + d.pos);
    let oldPopper = select(selector + ' .popper#popr' + geneID)
    if (oldPopper.nodes().length > 0) oldPopper.remove();
    var popperD3 = select(selector)
               .append('div')
               .attr('class', 'popper col-lg-4 col-md-8 col-sm-10')
               .attr('id', 'popr' + geneID);
    var popperHTML = get_PopperHTML(d);
    // popper content
    popperD3.append('div')
             .attr('class', 'popper-content')
             .html(popperHTML);
    if (nonEmptyArray(d.pfam)) {
        var doms = new Set();
        d.pfam.forEach(d => {
            if (d.class && d.class != '') {
                doms.add(d.class)
            }
        })
        var colors = colors269;
        var palette = scaleOrdinal()
                        .domain(doms)
                        .range(colors);
        protDomains(selector + ' #dom' + cleanString(d.anchor + d.pos),
                         d.pfam,
                         d.length || Math.abs((+d.end) - (+d.start)) || 1000,
                         250,
                         7,
                         palette,
                         URLs.pfam.b)
    }
    // Popper arrow
    popperD3.append('div')
             .attr('class', 'popper-arrow');

    var popper  = document.querySelector(selector + ' .popper#popr' + geneID);
    function show() {
        var poppers = document.querySelectorAll(selector + ' .popper')
        poppers.forEach(p => {
            p.removeAttribute('data-show');
        });
        let popper  = document
            .querySelector(selector + ' .popper#popr' + geneID);
        let ref = document
            .querySelector(selector + ' g.gene#gene' + geneID);
        popper.setAttribute('data-show', '');
        createPopper(ref, popper, {
          modifiers: [
            {
              name: 'offset',
              options: {
                offset: [-4, 5],
              },
            },
              {
                  name: 'flip',
                  options: {
                      fallbackPlacements: ['top'],
                  }
              }
          ],
        });
    }
    popper.addEventListener('click', show);
    return show;
}

var addPopper = function(selector,
                    id,
                    popperHTML,
                    popperClass) {
    var popperD3 = select(selector)
                    .append('div')
                    .attr('class', 'popper ' + popperClass)
                    .attr('id', 'popr' + id);
    // popper content
    popperD3.append('div')
            .attr('class', 'popper-content card-body h6 pt-2')
            .html(popperHTML);
    // popper arrow
    popperD3.append('div')
            .attr('class', 'popper-arrow');
    var popper = document.querySelector(selector + ' .popper#popr' + id);
    var ref = document.querySelector(selector + ' g#leaf' + id);
    function create() {
        // Popper Instance
        createPopper(ref, popper, {
          placement: 'right',
          modifiers: [
            {
              name: 'offset',
              options: {
                offset: [0, 10],
              },
            },
              {
                  name: 'flip',
                  options: {
                      fallbackPlacements: ['left'],
                  }
              }
          ],
        });
      }
      function show() {
        hide();
        popper.setAttribute('data-show', '');
        create();
      }
      function hide() {
        var poppers = document.querySelectorAll(selector + ' .popper')
        poppers.forEach(popper => {
            popper.removeAttribute('data-show');
        });
      }
      const showEvents = ['click'];
      showEvents.forEach(function (event) {
        popper.addEventListener(event, show);
        ref.addEventListener(event, show);
      });
}

var PopperClick = function(selector) {
    $(document).click(e => {
        // Helper function
        function lookForParent(element,
                             targetClass){
            let el = element;
            let name = el.nodeName;
            while (name && name != 'HTML') {
                if ($(el).hasClass(targetClass)) {
                    return el;
                }
                el = el.parentElement;
                name = el.nodeName;
            }
            return undefined;
        }
        let poppers = document.querySelectorAll(selector + ' .popper')
        poppers.forEach(popper => {
            popper.removeAttribute('data-show');
        });
        if (!e.altKey) {
            let targetID;
            ['gene', 'leaf', 'popper'].forEach(c => {
                try { targetID = lookForParent(e.target, c).id } catch {}
            })
            targetID = !targetID ? e.target.id : targetID;
            targetID = targetID.trim()
            if (['gene',  'leaf', 'popr'].indexOf(targetID.slice(0,4)) > -1){
                targetID = targetID.slice(4);
                let popper = document.querySelector(selector + ' .popper#popr'+targetID);
                //let popperDims = popper.getBoundingClientRect();
                //let refbound = document.querySelector(selector + ' g.gene#gene'+targetID)
                                       //.getBoundingClientRect();
                  //if (refbound.right+popperDims.width/2 > window.innerWidth){
                      //select(selector + ' .popper#popr'+targetID)
                          //.select(selector + ' .popper-arrow')
                          //.style('right', window.innerWidth-refbound.right+'px');
                  //} else if(refbound.left < popperDims.width/2) {
                      //select(selector + ' .popper#popr'+targetID)
                          //.select(selector + ' .popper-arrow')
                          //.style('left', refbound.left+'px')
                          //.style('right', '');
                  //}
                try { popper.setAttribute('data-show', '') } catch {}
            }
        }
    });
}

export {
    addPopper,
    PopperCreate,
    PopperClick
}
